docs/design/2023-11-29-priority-queue-for-auto-analyze.md

```diff
@@ -0,0 +1,270 @@
+[autoanalyze.go?L287:6]: https://sourcegraph.com/github.com/hi-rustin/tidb@d4618d4a5f91ee3703336fd1ba328c2e477652e5/-/blob/pkg/statistics/handle/autoanalyze/autoanalyze.go?L287:6&popover=pinned
+[autoanalyze.go?L301:40]: https://sourcegraph.com/github.com/hi-rustin/tidb@d4618d4a5f91ee3703336fd1ba328c2e477652e5/-/blob/pkg/statistics/handle/autoanalyze/autoanalyze.go?L301:40&popover=pinned
+[autoanalyze.go?L262:15]: https://sourcegraph.com/github.com/hi-rustin/tidb@d4618d4a5f91ee3703336fd1ba328c2e477652e5/-/blob/pkg/statistics/handle/autoanalyze/autoanalyze.go?L262:15&popover=pinned
+
+The above process is the current auto analyze process. We can see that the current auto analyze process is a random selection algorithm. We need to design a priority queue to solve the above problems.
+
+Since we perform analysis tasks synchronously on a single point for each table, we need to carry out weighted sorting on the tables that need analysis.
+
+   1. We still use tidb_auto_analyze_ratio to determine if the table needs to be analyzed. The default value is 0.5. It means that auto-analyze is triggered when greater than 50% of the rows in a table have been modified.
+   2. For special events, it is necessary to put the table into the priority queue. For example, the table has a new index but it hasn't been analyzed yet.
+   3. If a table is added to the auto-analysis queue, a weight must be assigned to it to determine the execution order.
+
+Weight table:
+| Name                 | Meaning                                                                 | Weight                                                                                                                                                                                                                                                                                                                                                            |
+|----------------------|-------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Percentage of Change | The percentage of change since the last analysis.                       | 50% <= x < 70%: 1 <br/> 70% <= x < 80%: 1.3 <br/> x >= 80%: 1.5  <br/>                                                                                                                                                                                                                                                                                            |
```

Comment:
> * Should these numbers be adjusted accordingly when `tidb_auto_analyze_ratio` changes?
> * Just curious, why do we choose to define 3 fixed weights instead of something like a linear function?

I used a fixed number because I wanted to start with a simple solution. We can always improve it later after testing. Perhaps using a linear function would be a better solution, but we can change it based on the results of our tests.

> 3\. For tables that are never analyzed, the percentage is 100%, right?

I want to consider this as a special event. However, it seems reasonable to assume that its percentage of change is 100% because all the data is new.

Reply:
> I used a fixed number because I wanted to start with a simple solution. We can always improve it later after testing. Perhaps using a linear function would be a better solution, but we can change it based on the results of our tests.

Got it. I have no problem with this.
So, do we need a number for `x < 50%` when the user changes `tidb_auto_analyze_ratio` to a smaller values?

> I want to consider this as a special event.

I think, according to the formula, even if it's a "Special Event", it still needs a `change_percentage` score to calculate the total score.